modules/params/keeper/keeper.go

```diff
@@ -1,51 +1,56 @@
 package keeper
 
 import (
-	"fmt"
-
-	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-
-	sdk "github.com/cosmos/cosmos-sdk/types"
-	paramskeeper "github.com/cosmos/cosmos-sdk/x/params/keeper"
-
 	"github.com/bianjieai/iritamod/modules/params/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
 // Keeper define a slashing keeper
 type Keeper struct {
-	paramskeeper.Keeper
+	authKeeper types.AccountKeeper
+
+	router *baseapp.MsgServiceRouter
 }
 
 // NewKeeper creates a slashing keeper
-func NewKeeper(keeper paramskeeper.Keeper) Keeper {
+func NewKeeper(ak types.AccountKeeper, router *baseapp.MsgServiceRouter) Keeper {
 	return Keeper{
-		keeper,
+		authKeeper: ak,
+		router:     router,
 	}
 }
 
-// HandleValidatorSignature handles a validator signature, must be called once per validator per block.
-// Block all subsequent logic if this validator has been removed.
-func (k Keeper) UpdateParams(ctx sdk.Context, params []types.ParamChange) ([]sdk.Attribute, error) {
-	var changeEvents []sdk.Attribute
-	for _, c := range params {
-		ss, ok := k.GetSubspace(c.Subspace)
-		if !ok {
-			return nil, sdkerrors.Wrap(types.ErrUnknownSubspace, c.Subspace)
+// UpdateParams updates the params
+func (k Keeper) UpdateParams(ctx sdk.Context, messages []sdk.Msg) error {
+	var (
+		err    error
+		events sdk.Events
+	)
+
+	cacheCtx, writeCache := ctx.CacheContext()
+	for _, msg := range messages {
+		handler := k.router.Handler(msg)
+		if handler == nil {
+			return sdkerrors.Wrap(types.ErrUnroutableUpdateParamsMsg, sdk.MsgTypeURL(msg))
 		}
 
-		k.Logger(ctx).Info(
-			fmt.Sprintf("attempt to set new parameter value; key: %s, value: %s", c.Key, c.Value),
-		)
+		var res *sdk.Result
 
-		if !ss.Has(ctx, []byte(c.Key)) {
-			return nil, sdkerrors.Wrapf(types.ErrUnknownKey, c.Key)
+		res, err = handler(cacheCtx, msg)
+		if err != nil {
+			break
```

Comment:
There's an appropriate way to solve this imo. Define a router wrapper `ParamsRouter` for `msgServiceRouter` in `iritamod/params`, which must implements two methods:

```go
type ParamsRouter interface {
  IsParamsType(msg sdk.Msg) bool
  Handler(msg sdk.Msg) (baseapp.MsgServiceHandler, bool)
}

func (pr *MockParamsRouter) IsParamsType(msg sdk.Msg) bool {
	switch msg.(type) {
	case *opbtypes.MsgUpdateParams:
		return true
        ...
        }

	return false
}
```

 By this way we can just add the approved update params msg type to `IsParamsType`, in our `irita` app. It would just reuse the `app.msgServiceRouter` in that register a new router which contians only update params msg type seem to be more compliticated in implementation.

Reply:
There is no need to define an interface, an example is as follows:

```golang 
type router struct {
   handlers map[string]sdk.MsgServiceHandler
}

func NewRouter(allowed []sdk.Msg,serverRouter,serviceRouter sdk.MsgServiceRouter) {
      handlers := make(map[string]sdk.MsgServiceHandler)
      for _, msg := range allowed {
           msgType := sdk.MsgTypeURL(msg)
           handler := serviceRouter.HandlerByTypeURL(msgType)
           if handler == nil {
               panic("no handler for msg")
           }
           handlers[msgType] = handler
       }
       return router{handlers}
}
```